nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/test/java/org/apache/nifi/processors/gcp/pubsub/PublishGCPubSubTest.java

```diff
@@ -0,0 +1,230 @@
+        runner.assertValid();
+
+        runner.setProperty(PublishGCPubSub.BATCH_SIZE_THRESHOLD, "-1");
+        runner.assertNotValid();
+        runner.setProperty(PublishGCPubSub.BATCH_SIZE_THRESHOLD, "15");
+        runner.assertValid();
+
+        runner.setProperty(PublishGCPubSub.BATCH_BYTES_THRESHOLD, "3");
+        runner.assertNotValid();
+        runner.setProperty(PublishGCPubSub.BATCH_BYTES_THRESHOLD, "3 MB");
+        runner.assertValid();
+
+        runner.setProperty(PublishGCPubSub.BATCH_DELAY_THRESHOLD, "100");
+        runner.assertNotValid();
+        runner.setProperty(PublishGCPubSub.BATCH_DELAY_THRESHOLD, "100 millis");
+        runner.assertValid();
+    }
+
+    @Test
+    void testSendOne_Success_FlowFileStrategy() throws InitializationException {
+        runner.setProperty(PublishGCPubSub.GCP_CREDENTIALS_PROVIDER_SERVICE, getCredentialsServiceId(runner));
+        runner.setProperty(PublishGCPubSub.TOPIC_NAME, "my-topic");
+        runner.setProperty(PublishGCPubSub.PROJECT_ID, "my-project");
+
+        runner.enqueue("text");
+        runner.run(1, true, true);
+        runner.assertAllFlowFilesTransferred(PublishGCPubSub.REL_SUCCESS, 1);
+        final MockFlowFile flowFile = runner.getFlowFilesForRelationship(PublishGCPubSub.REL_SUCCESS).iterator().next();
+        assertNotNull(flowFile.getAttribute(PubSubAttributes.MESSAGE_ID_ATTRIBUTE));
+    }
+
+    @Test
+    void testSendOne_Retry_FlowFileStrategy() throws InitializationException {
+        throwable = new UnavailableException(null, GrpcStatusCode.of(Status.Code.UNAVAILABLE), true);
+
+        runner.setProperty(PublishGCPubSub.GCP_CREDENTIALS_PROVIDER_SERVICE, getCredentialsServiceId(runner));
+        runner.setProperty(PublishGCPubSub.TOPIC_NAME, "my-topic");
+        runner.setProperty(PublishGCPubSub.PROJECT_ID, "my-project");
+
+        runner.enqueue("text");
+        runner.run(1, true, true);
+        runner.assertAllFlowFilesTransferred(PublishGCPubSub.REL_RETRY, 1);
+    }
+
+
+    @Test
+    void testSendOne_Failure_FlowFileStrategy() throws InitializationException {
+        runner.setProperty(PublishGCPubSub.GCP_CREDENTIALS_PROVIDER_SERVICE, getCredentialsServiceId(runner));
+        runner.setProperty(PublishGCPubSub.TOPIC_NAME, "my-topic");
+        runner.setProperty(PublishGCPubSub.PROJECT_ID, "my-project");
+        runner.setProperty(PublishGCPubSub.MAX_MESSAGE_SIZE, "16 B");
+        runner.enqueue("some really long text");
+
+        runner.run(1, true, true);
+        runner.assertAllFlowFilesTransferred(PublishGCPubSub.REL_FAILURE, 1);
+    }
+
+    @Test
+    void testSendOne_Success_RecordStrategy() throws InitializationException, IOException {
+        runner.setProperty(PublishGCPubSub.GCP_CREDENTIALS_PROVIDER_SERVICE, getCredentialsServiceId(runner));
+        runner.setProperty(PublishGCPubSub.TOPIC_NAME, "my-topic");
+        runner.setProperty(PublishGCPubSub.PROJECT_ID, "my-project");
```

Comment:
These initial `setProperty()` calls repeat the same values, so they could be pulled out to a private method that could be called when needed in each test method.